server/src/main/java/org/elasticsearch/common/lucene/store/BytesReferenceIndexInput.java

```diff
@@ -0,0 +1,128 @@
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+
+import java.io.EOFException;
+import java.io.IOException;
+
+public class BytesReferenceIndexInput extends IndexInput {
+
+    private final BytesReference bytesReference;
+
+    private int filePointer;
+    private StreamInput streamInput;
+
+    protected BytesReferenceIndexInput(String resourceDescription, BytesReference bytesReference) {
+        super(resourceDescription);
+        this.bytesReference = bytesReference;
+    }
+
+    @Override
+    public void close() throws IOException {
+    }
+
+    @Override
+    public long getFilePointer() {
+        return filePointer;
+    }
+
+    private StreamInput getOrOpenStreamInput() throws IOException {
+        if (streamInput == null) {
+            streamInput = bytesReference.slice(filePointer, bytesReference.length() - filePointer).streamInput();
+        }
+        return streamInput;
+    }
+
+    @Override
+    public void seek(long longPos) throws IOException {
+        if (longPos < 0) {
+            throw new IllegalArgumentException("Seeking to negative position: " + longPos);
+        } else if (longPos > bytesReference.length()) {
+            throw new EOFException("seek past EOF");
+        }
+        var pos = (int)longPos;
+        if (streamInput != null && filePointer <= pos) {
+            final var toSkip = pos - filePointer;
+            final var skipped = streamInput.skip(toSkip);
+            assert skipped == toSkip;
+        }
+        filePointer = pos;
+    }
+
+    @Override
+    public long length() {
+        return bytesReference.length();
+    }
+
+    @Override
+    public IndexInput slice(String sliceDescription, long offset, long length) throws IOException {
+        if (offset >= 0L && length >= 0L && offset + length <= bytesReference.length()) {
+            return new BytesReferenceIndexInput(sliceDescription, bytesReference.slice((int)offset, (int)length));
+        } else {
+            throw new IllegalArgumentException(
+                "slice() "
+                + sliceDescription
+                + " out of bounds: offset="
+                + offset
+                + ",length="
+                + length
+                + ",fileLength="
+                + bytesReference.length()
+                + ": "
+                + this
+            );
+        }
+    }
+
+    @Override
+    public byte readByte() throws IOException {
+        try {
+            return getOrOpenStreamInput().readByte();
+        } finally {
+            filePointer += 1;
+        }
+    }
+
+    @Override
+    public void readBytes(byte[] b, int offset, int len) throws IOException {
+        getOrOpenStreamInput().readBytes(b, offset, len);
+        filePointer += len;
+    }
+
+    @Override
+    public short readShort() throws IOException {
+        try {
+            return Short.reverseBytes(getOrOpenStreamInput().readShort());
```

Comment:
Do we even need to be this tricky with the stream here? 
Except for `short` we already have `getLongLE` and `getIntLE`  on the reference interface. Maybe we could do this simpler by just adding a version of `getShortLE` (or doing it manually here) and implementing the `readBytes` via the bytes ref iterator?
Since we track the `filePointer` manually, using the stream seems to just add complexity to this?

Reply:
The reason I'm using the stream is that it avoids the need to keep descending into a tree of `CompositeBytesReference` things when doing sequential reads.

Reply:
Ah I see, this stuff really isn't all that efficient for composite refs indeed ...